### DISCUSSION TITLE
Whether webGPU can render Yuv data directly to canvas

### DISCUSSION THREAD

**Discussion by yangfangfang1204** [gpuweb/gpuweb#2339] [Q&A]
i decode the video to the yuv data by wasm ， and next i want to render the yuv data to canvas by WebGPU, so is it supported ? and is there some demo ?

**Comment by shaoboyan** [comment]
@yangfangfang1204  My understanding is that you've already got YUV datas in multiple `ArrayBufferView` and want to use WebGPU to do post-processing, right?
So my suggestion is that:
- Use [`writeTexture`](https://gpuweb.github.io/gpuweb/#dom-gpuqueue-writetexture) or [`writeBuffer`](https://gpuweb.github.io/gpuweb/#dom-gpuqueue-writebuffer) to upload your contents from `ArrayBufferView` to `GPUTexture` or `GPUBuffer`
- Start `ComputePipeline/RenderPipeline` to do the post-processing. 

I believe you've got all the info that you need to do YCbCr->RGB conversion in wgsl.

Demo about using [`writeBuffer`](https://github.com/austinEng/webgpu-samples/blob/d2c0683e4ac74b99d1c24623c6407af1ce97466f/src/sample/imageBlur/main.ts);

**Comment by yangfangfang1204** [comment]
Thanks for your reply . my question is not post-processing , just display YUV data on canvas . so is there some demo ?

**Comment by shaoboyan** [comment]
@yangfangfang1204 Sry but I don't know there is online demos. And I believe the steps are the same for rendering to canvas.
You need to do YCbCr->RGB conversion in your wgsl shaders.

If your video source could be presented by `HTMLVideoElement` or could be rendered to `HTMLCanvasElement`, then [`importExternalTexture`](https://gpuweb.github.io/gpuweb/#dom-gpudevice-importexternaltexture) or [`copyExternalImageToTexture`](https://gpuweb.github.io/gpuweb/#dom-gpuqueue-copyexternalimagetotexture) could help.

Sample to use `importExternalTexture` with `HTMLVideoElement`: [https://github.com/austinEng/webgpu-samples/blob/71d987b5b1a1bf8accaf9aa0ff431f60e2319a84/src/pages/samples/videoUploading.ts](https://github.com/austinEng/webgpu-samples/blob/71d987b5b1a1bf8accaf9aa0ff431f60e2319a84/src/pages/samples/videoUploading.ts)

But you still need to write some wgsl shader to render the content to canvas.

**Comment by luojinghui** [comment]
There is a piece of code that uses WebGL and WebGPU to render YUV respectively, but there is an exception in WebGPU rendering. I can't see where there is a mistake? Can help locate the problem?

https://codepen.io/jkqbnbyh-the-flexboxer/pen/GRwoxwK?editors=0010

**Reply by kainino0x** [reply]
The browser Dev Tools console shows a bunch of WGSL syntax errors. You'll need to update to the current WGSL syntax. (`@vertex`, `@group`, `@binding`, etc)

```
Tint WGSL reader failure: :2:9 error: unexpected token
        [[stage(vertex)]]
...
```
